KEP-4444: Routing Preference for Services

gauravkghildiyal opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Adding new KEP-4444 proposing Routing Preference for Services

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4444

<!-- other comments or additional information -->
- Other comments: None

/sig network
/assign @thockin 
/cc @robscott, @aojea, @danwinship

gauravkghildiyal commented:
I will soon fill in the remainder of the sections but wanted to get the conversation started :)

gauravkghildiyal commented:
/cc @robscott
/cc @aojea 
/cc @danwinship

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+  strategy.
+* `PreferEqualSpread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
+* `PreferZone`: Encourages routing traffic to endpoints within the same zone as
```
Since this particular approach is aiming to be generic, I'd aim for something more like `PreferClose`, so that the same value could eventually be expanded to included region or other considerations as necessary.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+* This early mechanism allowed users to specify ordered routing preferences (e.g.,
+prioritize local nodes, then zone, then anywhere). 
+
+* **Limitations:** However, it lacked flexibility for implementations to make smart
+  decisions on its own, like incorporating feedback-based routing optimizations
+  (e.g., avoiding overloaded endpoints). The name `topologyKeys` might be
+  perceived as overly restrictive, given the desire to incorporate non-topology
+  factors into routing decisions.
```
Another key thing to mention here is the arbitrary ordering, lengths, and keys here made it very difficult to integrate with systems that might have a concept of more standard settings like preferring same zone routing. With topologyKeys you could theoretically request that an entirely arbitrary topology key be given preference over zone or region.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+  * Exposes the annotation service.kubernetes.io/topology-mode. When this
+    annotation is set to Auto, an implementation specific heuristic is used to
+    route the traffic. 
+  * **Goal:** The aim with Auto was to allow implementations to be as smart as
+    possible and choose what it considered to be the best routing criteria. The
+    field didn’t support any other value by design with the thought that the
+    implementation should be able to make the best possible choice for the user.
```
Maybe mention that this tried to find a balance between safety and same-zone routing, and unfortunately did not achieve either very well.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+    possible and choose what it considered to be the best routing criteria. The
+    field didn’t support any other value by design with the thought that the
+    implementation should be able to make the best possible choice for the user.
+  * **Limitation:** While designed for intelligent routing, this mode offers less
+    user control and can be less predictable. Some users prioritize
+    predictability for performance optimization, even when accepting potential
+    endpoint overload in certain situations.
```
I'd mention that multiple users have filed issues asking why hints weren't applied or working for them. More recent versions added events to help explain what was happening, but there was still a lot of confusion around the feature.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+    same node.
+  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.
+
+Note that while the initial proposal of InternalTrafficPolicy proposed a
+PreferLocal policy, it was dropped later on. This meant that now
+TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
+allow users to express a much desired use case from topologyKeys which is
+"prefer node-local, failover to same zone, then route anywhere"
```
Although this is all 100% correct, I don't think this KEP really addresses that?

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `PreferEqualSpread`: Encourages an equal distribution of traffic across
```
I know the name of the field is still up in the air, but we may need not need the `Prefer` prefix if the field name includes `Preference` as currently proposed.

thockin replied:
Also agree to drop "Prefer" from the values if it is in the field name

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+* **Solution:** Set `routingPreference=Default` (or leave the field unset)
+* **Effect:** The Kubernetes implementation will apply its best-effort routing
+  strategy based on its design. This strategy might change over time as the
+  implementation evolves. It may load balance across zones or regions regions.
```
```suggestion
  implementation evolves. It may load balance across zones or regions.
```

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+* **Requirement:** I prioritize application availability and want to minimize the
+  risk of outages due to localized overload. I'm willing to accept potentially
+  higher costs associated with cross-zone traffic distribution.
+* **Solution:** Set `routingPreference=PreferSpread`
```
Nit: This name doesn't match the name you have above.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+  traffic costs.
+
+#### Story 4
+* **Requirement:** As a developer deploying applications across multiple
+  Kubernetes environments, I want a consistent way to express my routing
+  preferences. 
```
I think what you're going for here is to describe the maintainer of what I'd call a cluster addon. For example, cert-manager, prometheus or really anything on https://artifacthub.io/. Maybe rephrase this to the following:
```suggestion
* **Requirement:** As a developer of a widely deployed cluster addon, I want to be
  able to provide users an easy way to configure my Helm chart and/or deployment
  configuration to enable same-zone routing behavior in a portable way that works
  across many different environments.
```

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+
+kube-proxy based dataplane (along with EndpointSlice controller, within
+kube-controller-manager as the control plane) will support the three standard
+routing preferences (`Default`, `PreferEqualSpread`, `PreferZone`).
```
Maybe worth starting a thread specifically to talk about the names of the values here. Although each of these make sense, I think I'd prefer names that were even a bit more generic, maybe:

- `Default`
- `Even`
- `Close`

This assumes that the name of the field already includes `Preference`. My main goal here would be to move away from `PreferZone` towards something that just allowed for `Close` routing, something that could include region or other keys in the future.

/cc @thockin @danwinship @aojea

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy based dataplane (along with EndpointSlice controller, within
```
```suggestion
kube-proxy (along with EndpointSlice controller, within
```

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+* This leverages existing implementation, requiring no major changes.
+
+#### `PreferZone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
```
Somewhere in this KEP we probably want to mention how we're going to change the existing kube-proxy logic around handling hints: https://github.com/kubernetes/kubernetes/blob/2363cdcc399cbf428210efb2c51575ddcad2b84a/pkg/proxy/topology.go#L151-L155. 

I tend to think that we should remove any check for an annotation or field on the Service and instead honor hints if _every_ endpoint for that Service has a hint specified.

gauravkghildiyal replied:
Oh good catch! 
For some reason I was under the impression that we had already made that change but turns out we haven't. I've added a section for this.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+EndpointSlice controller (which is acting as the control plane) will update the
+Service status with the following conditions (inspired by Gateway API conditions)
+
+* RoutingPreferenceAccepted
```
I think the only relevant component that could reliably set status on a Service here is the EndpointSlice controller. Given that, I think these are the states we'd want to be able to represent:

1. The EndpointSlice controller sees a value that is not domain-prefixed on a Service and does not know how to handle it
2. The EndpointSlice controller sees a value that it recognizes but can't populate hints due to some other factor (common with the current proportional approach)

It is likely worth representing the positive polarity versions of those as well:
1. The EndpointSlice controller sees a value that it recognizes and has populated hints or determined that it does not need to for the heuristic (Default and PreferEqualSpread would fall into the latter category as proposed)

I think the first item in both of those lists could be handled by this `RoutingPreferenceAccepted` condition that was true or false. Maybe it should be set to "Unknown" for unrecognized/domain-prefixed values?

I think the second item (error populating hints) should probably only appear in the error case. I think your suggestion of `RoutingPreferenceProgrammed` is reasonable here, but since this is very specific to hints, we could instead use something more specific like `HintsPopulated`.

gauravkghildiyal replied:
Thanks for giving this a detailed thought, Rob! I mostly have the same thoughts as you on the matter, with the exception of the topic regarding `RoutingPreferenceProgrammed` v/s `HintsPopulated` as the condition name. I was thinking that "HintsPopulated" would be a _reason_ under the `RoutingPreferenceProgrammed` condition. Here's what I had in mind:

* From the user's perspective, the top priority is whether their routing preference has been "somewhat" sent to the data plane. `RoutingPreferenceProgrammed` reflects this directly, while the use of hints may be thought of as an implementation detail (which we will convey through the condition's reason)

* Another reason might spur the debate of being generic vs _YAGNI_ :). But given the previous advantage, in this case I was leaning towards being generic which allows non-hint based implementations to reuse the same condition.

robscott replied:
Yep, that makes sense, I think that's a reasonable approach here.

robscott replied:
I think the primary reason we needed status on the previous iteration of topology was that there was so much confusion about when it was or was not applied. I'd recommend scrapping status from this stage of the KEP. We can choose to add it in the future if feedback suggests it would be helpful, but I'm hopeful that this will end up being closer to any other Service field where it's not really necessary.

robscott reviewed (commented):
Thanks for the work on this @gauravkghildiyal!

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+prevent conflicts when multiple control planes (like the default EndpointSlice
+controller) are present.
+
+### Choice of field name
```
Naming debate thread:

I don't love it.  "routing" feels really low-level.  We know this is about locality in practice, but maybe that is too concrete?  I guess, in theory, someone could use different criteria.

readers: ideas welcome

gauravkghildiyal replied:
How do folks feel about `trafficDistribution` with `Prefer` keyword in the field values? (Assuming `trafficDistributionPreference` might be too long)

danwinship replied:
agree that "routing" implies the wrong thing

> with `Prefer` keyword in the field values

moving "prefer" from the field name to the values opens up the possibility of having `Require` (or at least "stronger-than-Prefer") values in the future

thockin replied:
Final names can be argued in impl, but we will need to come back and update this.

sftim replied:
My tip: come up with really clear name for the feature gate, one that communicates what the feature is and does. Then see if that helps us name the API fields.

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[ ] I/we understand the owners of the involved components may require updates to
```
Gotta check the box

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,1128 @@
+-->
+
+### Upgrade / Downgrade Strategy
+
```
This section will be of interest

thockin reviewed (commented):
I recognize that this is a first draft, but it looks pretty good (though I got a sneak peak, so most of my concerns are addressed).  

Please put some energy into the operational and PRR parts, next.

gauravkghildiyal commented:
Hi @johnbelamaric -- would you be able to help with the PRR review for this KEP? The target is alpha release for the coming 1.30 version (Yes the freeze is just around the corner :) )

gauravkghildiyal commented:
/assign @johnbelamaric for PRR

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -339,34 +295,37 @@ NOTE: Implementations reserve the right to refine the behavior associated with
-Go in to as much detail as necessary here.
-This might be a good place to talk about core concepts and how they relate.
--->
+None.
```
Caveat: This being our third attempt at an API for this, there's a non-zero chance that we STILL didn't get it right, and will have to revisit again.

Caveat: Converting from the annotation to the field is not quite a mechanical change, which means users will have to think about it.

gauravkghildiyal replied:
Added 1.
2 is part of Risks and Mitigations.

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -339,34 +295,37 @@ NOTE: Implementations reserve the right to refine the behavior associated with
+  annotation might encounter differences in exact routing behavior:
+  * The new field doesn't support a routing preference that is exactly similar
+    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.
+  * If both field and the annotation are set, the annotation will take precedence.
```
...for some number of releases, until it is deprecated and removed

gauravkghildiyal replied:
Added

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -405,6 +364,23 @@ NOTE: The fact that EndpointSlice hints are not expected to be implemented by
+**New behaviour:** Irrespective of what the annotation
+`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
+set to (or even if they are not set at all), kube-proxy will always consider
+EndpointSlice hints.
```
"... kube-proxy will always consider EndpointSlice hints (assuming this feature gate is enabled)"

gauravkghildiyal replied:
Added!

thockin reviewed (commented):
/lgtm
/approve

merge and iterate - names are the big outstanding

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,908 @@
+-->
+
+## Alternatives
+
```
Not sure that folks are going to like it, but: I'd like to see an alternative discussed where the appropriate heuristic to use is captured in a new API kind.

Let's call this TargetPodHeuristic for now.

So a Service might look like:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: example-service
  namespace: default
spec:
  heuristics:
    - example-heuristic
  selector:
    app.kubernetes.io/name: example
  ports:
    - protocol: TCP
      port: 24
      targetPort: 24
```

and the TargetPodHeuristic for that:
```yaml
apiVersion: example/v1alpha1
kind: TargetPodHeuristic
metadata:
  name: example-heuristic
  namespace: default
spec:
    preferredTargets:
      - weight: 23
        matchLabelKeys:
         - topology.kubernetes.io/zone   # prefer a target in the same zone
         - topology.kubernetes.io/region # prefer a target in the same region
                                         # maybe zone names are only regionally unique
      - weight: 11
        matchLabelKeys:
         - topology.kubernetes.io/region # prefer a target in the same region
    avoidedTargets:
      - weight: 42
        matchTaints:
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoSchedule"
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoExecute"
```

To be clear: we can rule this out if we want to. That's absolutely fine. What I'm calling for is enough of a write-up about why it wouldn't fit / couldn't work / doesn't scale / etc. Let's not leave readers to guess this themselves.

gauravkghildiyal replied:
sftim -- That's a reasonable ask. Thanks. I've tried to add something. Hope that conveys the right ideas.

thockin replied:
This skews back to `topologyKeys` which gave users a ton of control but made alternative implementations very difficult because it was over-precise.

It's a fine line we are walking between defining a full service mesh API and defining too little to be useful.

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `Spread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
```
That sounds like the behavior that is currently expected from kube-proxy (EDIT: yes, basically confirmed below), and yet, by implication, it is _not_ what `Default` means? It seems to me that in past discussions Tim was pretty insistent that kube-proxy _must_ "encourage an equal distribution of traffic across endpoints" by default.

So maybe this should be the `Default`/`nil` behavior, and you should have `Auto` (?) for the "no specific routing preference" behavior.

robscott replied:
I think the idea here is to not completely prevent us from having smarter default behavior in the future, but I agree that the default in most cases should be `Spread`. I'd just like to hold out hope that somehow there can be a sufficiently safe "Close" mode that could be enabled by default in the future.

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  `Zone` heuristic – the hints will match the zone of the endpoint itself.
+* While it may seem redundant to populate the hints here since kube-proxy can
+  already derive the zone hint from the endpoints zone (as they would be the
+  same), we will still use this for implementation simply because of the reason
+  that it’s easier to implement and provides a better design. Consider an
+  alternative implementation where kube-proxy reads
+  `routingPreference=Zone` field and then constructs the route rules
+  accordingly. This means some extra logic needs to be baked into the kube-proxy
+  which could have just as easily been implemented by an already existing
+  extensibility mechanism (i.e. EndpointSlice hints)
```
I don't understand this. It sounds like you are saying "having two fields that will always have the same value is better design than only having one"

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+
+#### `Zone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  `Zone` heuristic – the hints will match the zone of the endpoint itself.
```
So how will `routingPreference: Zone` address the aforementioned Limitations of Topology-Aware Hints?

robscott replied:
I think the biggest benefit of this simpler approach is the predictability and simplicity of this approach. It will also work equally well for Services with small numbers of endpoints. If I were to summarize the feedback I've gotten about the current approach, it would largely fall into some combination of the following:

1. Why aren't hints being set?
2. How do I ensure that hints will be reliably set?

This approach should avoid both of those common points of frustration.

danwinship replied:
> It seems like CPU-proportional is a more or less failed experiment.

Oh... I see. When the KEP says "the hints will match the zone of the endpoint itself", it's talking about _making a change_ to kcm, not explaining the current behavior of kcm. That is not clear (and is the source of my confusion below as well).

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+**New behaviour:** Irrespective of what the annotation
+`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
+set to (or even if they are not set at all), kube-proxy will always consider
+EndpointSlice hints (assuming this feature-gate is enabled).
```
It will "consider" hints regardless of `routingPreference` value? That needs clarification. I assume it "considers" them in very different ways for `Spread` vs `Zone`.

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+  trafficRoutingPreference was not favored so that it’s clearly different from
+  the `internalTrafficPolicy` and `externalTrafficPolicy` fields which have a
+  more fixed behavior.
+* Use of the word "selection" for a name like `endpointSelection` were avoided
+  so as not to confuse with the actual process of selecting the complete set of
+  pods backing a service.
```
`endpointSelection` was my suggestion before I got to this comment... I'm not sure if I agree or disagree with that objection...

maybe `endpointDistribution`?

danwinship commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```
At some point in the last discussion, we discovered [`pod.spec.topologySpreadConstraints`](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/), which are a way of saying things like "I want this Deployment deployed such that each zone/node/whatever has the same number of Pods", and there was [a brief discussion](https://github.com/kubernetes/enhancements/pull/3293#issuecomment-1322698198) about how we could use that for topology-aware service routing.

robscott replied:
+1 - definitely worth covering that option in this section. My personal opinion is that that API is both too complex for users and too limiting for implementations.

gauravkghildiyal replied:
Thanks for the suggestions. I've added a section on `Reuse Pod Topology Spread Constraints for Traffic Distribution`. Additionally, feels like rather than being an either-or situation, it's more appropriate for Pod Topology Spread Constraints to complement the `routingPreference`. Added a section on `Complementary Use of Pod Topology Spread Constraints and routingPreference` which eventually might also make sense in a user facing doc

> My personal opinion is that that API is both too complex for users and too limiting for implementations.

I share the same thoughts. The section on `Granular Routing Controls` below exactly emphasis those downsides.

gauravkghildiyal commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+    same node.
+  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.
+
+Note that while the initial proposal of InternalTrafficPolicy proposed a
+Local policy, it was dropped later on. This meant that now
```
Thanks for clarifying that, Rob. Yes that was the idea (seems like I mistakenly did a find replace in a some commit :) ). Will fix this.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+
+The field will support the following initial values:
+
+* `Default`: Indicates no specific routing preference. The user delegates the
```
I don't have a strong opinion here, but can agree with just letting a `nil` value here represent the default lack of preference.

thockin replied:
This could be haggled in the imp[lementation, too.  Need not block the KEP

thockin replied:
Can you mark in UNRESOLVED?

danwinship replied:
> Perhaps "Implementation" or "None" is better?

I mean, "None" feels even more like it should be `nil`. If you have a preference about how to route, you specify `routingPreference: Foo`. If you don't have a preference, you just don't specify anything.

robscott replied:
> If you don't have a preference, you just don't specify anything

I think we're all agreeing that we don't want users to have to specify anything in this case, the real question is if there's value in us setting a default value here that would show up on reads + be easier to revert to than nil.

thockin replied:
Let's discuss in impl, but I could be OK with the default being nothing (note: "" is something :)

gauravkghildiyal reviewed (commented):
Added section on Pod Topology Spread Constraints. Fixed some typos.

gauravkghildiyal commented on keps/sig-network/4444-service-routing-preference/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+  - "@gauravkghildiyal"
+  - "@robscott"
+owning-sig: sig-network
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
```
Thanks!

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,945 @@
+  strategy.
+* `Spread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
+* `Zone`: Encourages routing traffic to endpoints within the same zone as
```
I hate rehashing this, but I am still unsure.

I don't think we want to be too precise. Calling this "Zone" implies that "Region" and "Node" don't factor.  I don't think we want very many of these.  If "Zone" means anything significantly different than "use zone locality", the name is wrong or the implementation is wrong.

> Now, If we were to choose the heuristic name of Close right now (which at the moment would mean "route traffic within the same zone") and then later on we want to introduce an equivalent to Local (still without any feedback mechanism available to us), I don't think we would be able to say that, henceforth, Close means "route traffic within the same node" because that is increasing the risk. 

I don't think I agree.  If the new definition of "Close" means significantly more risk, the implementation should expose it with a new name.  But is that on US to define?  Adding a standard mode should have a VERY VERY high bar

danwinship replied:
I don't think anyone wants a mode that prefers-but-does-not-require same-zone, but which does not also prefer-but-not-require same-region and same-node.

It seems to me like there are three kinds of users/services:
1. "I don't have any special requirements, please just route my services efficiently, and if you can change the routing rules to make it even more efficient in the next release then yay"
2. "I don't have any special requirements, but I don't want to wait for you to make service routing more efficient (or I think I'm smarter than you), so please give me back my `topologyKeys`."
3. "I have specific requirements, which may involve money, and want to be able to enforce those requirements precisely, without worrying that a future version of kube-proxy will optimize around them"

If we aren't going to humor group 2, then groups 1 and 2 collapse together. They basically want "send to the endpoint that best balances (a) low latency, (b) high throughput, (c) the future availability of low-latency high-throughput connections for all clients". Though they can benefit from vaguely-defined routing modes, which will help them express hints like "prioritize latency" (`Close`) or "prioritize throughput/reliability" (`Spread`) so "dumb" proxy implementations can get slightly better behavior, while "smarter" proxy implementations can balance those vague hints against what their own endpoint monitoring is telling them.

But group 3 primarily wants to avoid being charged for cross-zone traffic. If their cluster is in a situation where zone 1's endpoints are getting overloaded and zone 2's endpoints are idle, they don't want Kubernetes to automatically start routing traffic to zone 2 for them, they want Kubernetes to automatically deploy more endpoints to zone 1 for them. So they want modes that are extremely precise about some things, though they can still be vague about other things. (eg, if they say `RequireSameZone`, the proxy is still allowed to try to optimize endpoint balance _after_ it has enforced that constraint).

If we don't care about group 3, then I don't think there's any reason for the API to mention zones at all. It doesn't make sense to have a mode that means "keep traffic within the same zone if it makes sense, but let it go to other zones if that would be better", because that should just be part of `Default`.

thockin replied:
We think almost everyone should choose Default, right?  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."  

The user-story around helm charts is what keeps coming back to my mind.  If I know that I have set up zonal scaling and so forth, it's OK for me to choose "fast" over "good" (because I handled Goodness myself).

So maybe the question is whether this is really a PREFERENCE that implementations may ignore (I know you said prefer zone, but I know from feedback that this is bad) or strict?  If you really really want strict same-zone, maybe you should have a different Service per zone?  That's crappy for LBs, I guess.

So I'll ask again: does "Zone" mean an implementation must ignore information and make "wrong" choices?  Does "Spread" mean it must cross zones sometimes, just to say it approached a random distribution?

danwinship replied:
> We think almost everyone should choose Default, right?

In the future, yes, almost everyone should choose Default (while they zip around Mars with their jetpacks). In the short run though, we know nobody is actually doing endpoint monitoring, and we don't have any specific plans to add it to kube-proxy, so hints seem useful. (But we want to define the hints so that future jetpack kubernetes isn't overly constrained by them.)

>  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."

zone is sometimes about "cheap", not just "fast"

> So maybe the question is whether this is really a PREFERENCE that implementations may ignore

There are people who at least claim that they want strict.

>  ("I know you said prefer zone, but I know from feedback that this is bad")

It was bad at the point you gathered the feedback, but what if their helm chart includes a HorizontalPodAutoscaler that is already in the process of fixing that?

We can assume that the service proxy is more clever than the average user, but we can't assume it's more clever than _every_ user...

---

Another (possibly implausible) "strict" example: "clients in the EU region must be routed to endpoints in the EU region, for GDPR compliance". If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections get _dropped_ than be misrouted. (Which is probably not actually what most people want in the "RequireZone" case...)

robscott replied:
> If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections fail than be misrouted. 

This is a good point and something that the current generation of hints are sadly unprepared to handle since we simply fall back to cluster-wide routing if no endpoints have matching zone hints. I think historically we've said that "RequireZone" and "RequireRegion" were sufficiently niche to be out of scope for this, but it's worth noting that not just are they out of scope, they'd require different implementation details (although of course it could just be new values added to this API in the future).

The very nature of the proposed field name here indicates that it's a preference and not a policy, similar conceptually to "hint" vs "requirement". I think that we might be able to make a case for something like "RequireZone" belonging as an option in internal and external traffic policy fields in the future instead of this preference field, but maybe that's an unnecessary and unhelpful distinction. In that world you'd have the following logic for each type of traffic:

_Is the policy field set to a non-"cluster" value for this type of traffic?_

Yes: Policy has precedence, policy is likely ignored unless it's somehow meaningful within the subset of endpoints the policy results in
No: Use the routing preference specified by this new field

Does that logic make sense? Is it weird to have external vs internal distinctions for traffic policies but not this routing preference?

For reference, what I've described above is pretty similar to [how we handle this today in kube-proxy](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/pkg/proxy/topology.go#L40) with hints, with the difference being that this leaves open the possibility that both a policy and a preference could both be applied for the same service which is not currently possible.

thockin commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  the client. If no endpoints are available within the zone, traffic should be
+  routed to other zones.
+
+Implementations are strongly encouraged to support the standard values. While
```
"Strongly encoraged" != required.

Is it OK for an implementation to not support "Zone" at all?  If so, should it accept the value and ignore it, or should it fail to program endpoints at all?

thockin reviewed (commented):
I'm OK to merge and iterate on names.  I think the biggest question is how optional the "strong suggestions" are and what implementations should do when they don't support the spirit of a standard name.

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+## Goals
+
+* **Enhanced User Control:** Provide users with a knob to specify a preference that
+  MAY influence the traffic routing of a Kubernetes Service.
```
```suggestion
  **may** influence the traffic routing of a Kubernetes Service.
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+## Non-Goals
+
+* **Strict Routing Guarantees:** The `routingPreference` field will not
+  enforce deterministic routing paths. It serves as a mechanism for expressing
+  hints and preferences that implementations can consider when making routing
+  decisions.
```
If the guarantees aren't there, are we as a side effect making it harder for Kubernetes to offer those guarantees one day?

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  interpretation might vary across implementations.
+
+* **Replacement of Traffic Policies:** The new field is complementary to
+  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
```
```suggestion
  `internalTrafficPolicy` and `externalTrafficPolicy`. It does not aim to substitute
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+* **Requirement:** I don't have strong preferences for how my application
+  traffic is routed. I prioritize simplicity and trust my Kubernetes
+  implementation to optimize traffic distribution.
+* **Solution:** Set `routingPreference=Default` (or leave the field unset)
```
(nit)
```suggestion
* **Solution:** Set `routingPreference: Default` (or leave the field unset)
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+* **Requirement:** I want my application to primarily receive traffic from
+  endpoints within the same zone for performance or cost reasons. However,
+  I want to avoid connection failures if no local endpoints are available.
+* **Solution:** Set `routingPreference=Zone`
```
(nit)
```suggestion
* **Solution:** Set `routingPreference: Zone`
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  endpoints within the same zone for performance or cost reasons. However,
+  I want to avoid connection failures if no local endpoints are available.
+* **Solution:** Set `routingPreference=Zone`
+* **Effect:** The Kubernetes implementation will aim to prioritize routing
+  traffic to endpoints in the same zone as the client. If no endpoints are
+  available within the zone, traffic will be routed to other zones. It's
+  possible that traffic patterns could lead to endpoints within the preferred
+  zone becoming overloaded. Consider other routing strategies or scaling up
+  resources within the zone if this becomes a concern.
```
For this user story, it sounds as if the user gets to find out if their preference worked once they see the next bill. I predict unhappy finance teams.

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+* **Requirement:** I prioritize application availability and want to minimize the
+  risk of outages due to localized overload. I'm willing to accept potentially
+  higher costs associated with cross-zone traffic distribution.
+* **Solution:** Set `routingPreference=Spread`
```
(nit)
```suggestion
* **Solution:** Set `routingPreference: Spread`
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+* **Requirement:** I have some other precise preferences for how traffic should
+  be routed, and I know that my chosen implementation supports the desired
+  preference. 
+* **Solution:** Set `routingPreference=<domain>/<heuristicName>` (where
```
(nit)
```suggestion
* **Solution:** Set `routingPreference: <domain>/<heuristicName>` (where
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  heuristic. It's important to note that the precise behavior of
+  implementation-specific heuristics might vary.
+
+### Notes/Constraints/Caveats (Optional)
```
(nit)
```suggestion
### Notes/Constraints/Caveats
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+#### `Zone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
```
(nit)
```suggestion
* We already use Hints to implement `service.kubernetes.io/topology-mode: Auto`
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  same), we will still use this for implementation simply because of the reason
+  that it’s easier to implement and provides a better design. Consider an
+  alternative implementation where kube-proxy reads
+  `routingPreference=Zone` field and then constructs the route rules
```
```suggestion
  `routingPreference: Zone` and then constructs the route rules
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+set to (or even if they are not set at all), kube-proxy will always consider
+EndpointSlice hints (assuming this feature-gate is enabled).
+
+NOTE: The expectation remains that *all* endpoints within an EndpointSlice must
+  have corresponding hints for kube-proxy to utilize them. This avoids scenarios
+  with partial hints. The reason for this requirement is the same one highlighted in [KEP-2433 Topology
+  Aware
+  Hints](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2433-topology-aware-hints/README.md#kube-proxy), i.e. _"This is to provide safer transitions between enabled and disabled states. Without this fallback, endpoints could easily get overloaded as hints were being added or removed from some EndpointSlices but had not yet propagated to all of them."_
```
Is it OK to have multiple EndpointSlices for one Service, where some of those EndpointSlices have hints for all endpoints and other EndpointSlices have no hints? And, if it is OK, is it OK for that to be a lasting situation?

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+future when this is possible, another condition like the following could be
+used:
+
+* RoutingPreferenceReady
+  * **Type:** `RoutingPreferenceReady`
+  * **Description:** Confirms that the dataplane has received the hints,
+    acknowledged them, and configured itself accordingly.
```
It'll be tricky to get right if we make that available - it requires aggregating the underlying status from lots of places.

---

I'd call this `RoutingPreferenceHonored`. Because it's a preference, not honoring it is OK - but people still want to know if it did get honored.

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+    <td>-</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>routingPreference=Auto</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>Local</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>InternalTrafficPolicy=Local</td>
```
```suggestion
    <td>routingPreference: Auto</td>
    <td>routingPreference: Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>-</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy: Local</td>
    <td>routingPreference: Auto</td>
  </tr>
  <tr>
    <td>Local</td>
    <td>Local</td>
    <td>Auto</td>
    <td>ExternalTrafficPolicy: Local</td>
    <td>InternalTrafficPolicy: Local</td>
```

This is valid YAML (snippets) and is closer to valid JSON than the version using `=`.

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+##### Integration tests
+
+* Verify that if both the annotation `service.kubernetes.io/topology-mode=Auto`
+  and field `routingPreference=Zone` are configured, precedence is given
+  to the annotation.
```
```suggestion
* Verify that if both the annotation `service.kubernetes.io/topology-mode: Auto`
  and field `routingPreference: Zone` are configured, precedence is given
  to the annotation.
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+##### e2e tests
+
+* Verify that EndpointSlice hints are correctly populated when
+  `routingPreference=Zone`
+* Verify through probes that when `routingPreference=Zone`, requests
```
(nit)
```suggestion
* Verify that EndpointSlice hints are correctly populated when
  `routingPreference` is set to `Zone`
* Verify through probes that for `routingPreference: Zone`, requests
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+* Verify that EndpointSlice hints are correctly populated when
+  `routingPreference=Zone`
+* Verify through probes that when `routingPreference=Zone`, requests
+  originating from a zone which has service pods get sent to a pod in the same
+  zone. For requests originating from zones with no service pods, requests
+  should not get blackholed and should rather be forwarded to any service pod
+  from the cluster.
```
Can we usefully end-to-end test a _preference_? It seems to me that the only thing we can actually assert is that traffic does not get dropped into a hole.

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+#### Alpha
+
+- Feature implemented behind a feature flag
```
(nit)
```suggestion
- Feature implemented behind a feature gate
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  then field `routingPreference`'s value will be assumed to be `Default` and the
+  behaviour will be the same as above with the annotation taking precedence.
+
+Upon downgrade of EndpointSlice controller (within kube-controller-manager) and
+kube-proxy, the `routingPreference` will simply not be considered in any
+decisions.
```
For downgrade, how would the API server handle the `routingPreference` field that it finds on some persisted objects? Any concerns there?

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+
+* **kube-apiserver:** [Kubernetes Version Skew
+  Policies](https://kubernetes.io/releases/version-skew-policy/#supported-version-skew)
+  require that kube-apiserver is atleast at the version of kube-proxy or
```
```suggestion
  require that kube-apiserver is at least at the version of kube-proxy or
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  kube-controller-manager. The only valid version skew would mean that a newer
+  kube-apiserver serves the new `routingPreference` field but the older
+  kube-proxy and kube-controller-manager would silently ignore this field. (No
+  adverse affect, behaviour equivalent to feature being disabled)
```
(nit)
```suggestion
  adverse affect, behaviour equivalent to feature being disabled).
```

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  pod distribution across zones, maximizing the likelihood that the
+  `routingPreference` can be satisfied and reduce (although not completely
+  eliminate) chances of overload for a single zone.
+
```
Is there an alternative available where we put the complexity into Gateway to ensure that there is a SomethingWeNeedToName (ie, not Service) per zone, and it does, and we're happy?

In this story:
- maybe a SomethingWeNeedToName controller manages the EndpointSlices, rather than kube-controller-manager
- maybe you can still refer to the Service if you want to, for example using [environment variables](https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables) - and the Cluster IP for that Service sends traffic somewhere reasonable.
- maybe we annotate Services when we want SomethingWeNeedToNames set up per zone.


That's not fleshed out; the key ask is: _did we already consider putting the complexity [mostly] out of tree into Gateway, with kube-proxy learning how to integrate and other in-tree code remaining unaware?_

sftim commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,998 @@
+  eliminate) chances of overload for a single zone.
+
+
+## Infrastructure Needed (Optional)
+
+N/A
```
```suggestion
```

sftim reviewed (commented):
Even more feedback. I hope it's useful.

BenTheElder reviewed (commented):
[PRR Shadow]
PRR LGTM for Alpha.

johnbelamaric commented:
Ok, I see Tim has approved for the SIG. The PRR looks fine for alpha. Note Ben's comment though.

/approve

thockin commented:
@gauravkghildiyal @danwinship @robscott and I chatted briefly yesterday - is this KEP updated to reflect that?

gauravkghildiyal commented:
> @gauravkghildiyal @danwinship @robscott and I chatted briefly yesterday - is this KEP updated to reflect that?

I will update in an hour and mark those things unresolved. Will revert back. Thanks for the discussion.

gauravkghildiyal commented:
@thockin, @danwinship, @robscott -- Ready for final review/approvals.

Cleaned things up as per our latest discussion. Summarizing our discussion and the accompanying deltas to this doc:

* We will start with having only two standard values: `Close` and `Default`
    * Changed doc to clarify what `Close` would mean for kube-proxy/endpointslice-controller, as of now (viz. Close == PreferZone)
* Changed “Implementations are strongly encouraged to support standard values” to “Implementations SHOULD support standard values”
   * This should allow use cases like `User Story 3` involving deployments through helm charts working across implementations.
   * The only standard value with some specific meaning defined is for `Close`, but the meaning still allows many implementations to be compliant. 
* The name of the field `routingPreference` is still unresolved and will most likely be refined during implementation.

robscott commented on keps/sig-network/4444-service-routing-preference/README.md:
```diff
@@ -0,0 +1,980 @@
+interpretation is permitted, implementations should aim to align their behavior
+with the described intent of these preferences as closely as possible.
+
+Implementations may support additional routing heuristics using values of the
+form `<domain>/<heuristicName>`. Heuristics without a domain prefix will be
+reserved for potential future standardization.
```
I'd personally recommend leaving this as an optional future enhancement to limit the scope of this KEP for now. We can always loosen validation to allow this in the future if there's demand for it.

robscott reviewed (approved):
Thanks @gauravkghildiyal! A couple non-blocking suggestions for where we can lessen the scope of this KEP but otherwise LGTM.

gauravkghildiyal commented:
Thanks @robscott -- incorporated the suggestions by adding an optional future expansions section.

robscott commented:
Thanks @gauravkghildiyal!

/lgtm
